core/src/test/scala/unit/kafka/server/TxnOffsetCommitRequestTest.scala

```diff
@@ -0,0 +1,133 @@
+  def testTxnOffsetCommitWithNewConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testTxnOffsetCommit(true)
+  }
+
+  @ClusterTest(
+    serverProperties = Array(
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+    )
+  )
+  def testTxnOffsetCommitWithOldConsumerGroupProtocolAndNewGroupCoordinator(): Unit = {
+    testTxnOffsetCommit(false)
+  }
+
+  @ClusterTest(
+    serverProperties = Array(
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.NEW_GROUP_COORDINATOR_ENABLE_CONFIG, value = "false"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.GROUP_COORDINATOR_REBALANCE_PROTOCOLS_CONFIG, value = "classic"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = GroupCoordinatorConfig.OFFSETS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_PARTITIONS_CONFIG, value = "1"),
+      new ClusterConfigProperty(key = TransactionLogConfig.TRANSACTIONS_TOPIC_REPLICATION_FACTOR_CONFIG, value = "1"),
+  ))
+  def testTxnOffsetCommitWithOldConsumerGroupProtocolAndOldGroupCoordinator(): Unit = {
+    testTxnOffsetCommit(false)
+  }
+
+  private def testTxnOffsetCommit(useNewProtocol: Boolean): Unit = {
+    if (useNewProtocol && !isNewGroupCoordinatorEnabled) {
+      fail("Cannot use the new protocol with the old group coordinator.")
+    }
+
+    val topic = "topic"
+    val partition = 0
+    val transactionalId = "txn"
+    val groupId = "group"
+
+    // Creates the __consumer_offsets topics because it won't be created automatically
+    // in this test because it does not use FindCoordinator API.
+    createOffsetsTopic()
+
+    // Join the consumer group. Note that we don't heartbeat here so we must use
+    // a session long enough for the duration of the test.
+    joinConsumerGroup(groupId, useNewProtocol)
+
+    var adminClient: Admin = null
+    var producer: Producer[String, String] = null
+
+    try {
+      adminClient = cluster.admin()
+      adminClient.createTopics(singleton(new NewTopic(topic, 1, 1.toShort)))
+
+      producer = cluster.producer(Collections.singletonMap(ProducerConfig.TRANSACTIONAL_ID_CONFIG, transactionalId))
+
+      producer.initTransactions()
+      producer.beginTransaction()
+      // add this line to avoid error INVALID_TXN_STATE
+      producer.sendOffsetsToTransaction(
+        Collections.singletonMap(new TopicPartition("topic", 0), new OffsetAndMetadata(5)),
+        new ConsumerGroupMetadata(groupId))
```

Comment:
I suppose that you must do this in order to add the __consumer_offsets partition to the transaction.

Reply:
Personally, I prefer to build the `AddOffsetsToTxnRequest` explicitly. for example:
```
  protected def addOffsetsToTxn(groupId: String,
                                producerId: Int,
                                producerEpoch: Short,
                                transactionalId: String,
                                version: Short): Unit = {
    val request = new AddOffsetsToTxnRequest.Builder(new AddOffsetsToTxnRequestData()
      .setTransactionalId(transactionalId)
      .setProducerId(producerId).setProducerEpoch(producerEpoch).setGroupId(groupId)
    ).build(version)

    val expectedResponse = new AddOffsetsToTxnResponseData()

    val response = connectAndReceive[AddOffsetsToTxnResponse](request)
    assertEquals(expectedResponse, response.data)
  }
```

Reply:
Thank you both! I believe all comments have been addressed. Please take a look when you have a chance.